### ISSUE TITLE
09-k-means-housing.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/proyecto-machine-learning-gregoriant#7]
Hola @gregangel1550-del ! el ejercicio tiene la estructura básica pero le falta bastante contenido. ¡Ánimo!
Lo más importante:

1. **Data leakage** al escalar antes del split
2. **No hay visualizaciones de clusters** - es fundamental ver cómo quedaron los grupos geográficamente
3. **Sobrescribes `model`** - pierdes el K-Means cuando creas el RandomForest
4. **No guardas los modelos**
5. **MUY IMPORTANTE: No hay conclusiones** ni interpretación de los resultados

Añade las visualizaciones, conclusiones y las métricas que faltan para completar el ejercicio.

## LO QUE HAS HECHO BIEN

- **Selección de variables correcta**: Latitude, Longitude, MedInc
- **Histogramas de distribución**: Visualización inicial de cada variable
- **MinMaxScaler**: Escalado correcto para K-Means
- **KElbowVisualizer**: Excelente uso de Yellowbrick para el método del codo
- **K-Means con k=6**: Valor correcto
- **Train/test split**: Correcto con test_size=0.2
- **RandomForestClassifier**: Modelo supervisado adecuado
- **Accuracy en DataFrame**: Buena presentación de train y test juntos

---

### 1. ERROR: Escalas TODO el dataset ANTES del split (DATA LEAKAGE)

```python
min_max_features = min_max_scaler.fit_transform(df[num_variables])  # ← Escalas TODO
df_min_max = pd.DataFrame(min_max_features, ...)
# ... luego split sobre datos ya escalados
```

El scaler "ve" los datos de test al calcular min/max. La forma correcta:

```python
# 1. Split PRIMERO
X_train, X_test = train_test_split(df, test_size=0.2, random_state=42)

# 2. Escalar DESPUÉS
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)  # Solo transform, no fit
```
# aqui puedes haccer el método del codo si quieres
---

### 2. Falta visualización de los clusters

No tienes ningún scatter plot de los clusters y como han quedado:

```python
fig, axes = plt.subplots(1, 3, figsize=(15, 5))

# Lon vs Lat
axes[0].scatter(df_min_max["Longitude"], df_min_max["Latitude"], 
                c=df_min_max["cluster"], cmap='viridis', alpha=0.6)
axes[0].set_title("Longitude vs Latitude")

# Lat vs MedInc
axes[1].scatter(df_min_max["Latitude"], df_min_max["MedInc"], 
                c=df_min_max["cluster"], cmap='viridis', alpha=0.6)
axes[1].set_title("Latitude vs MedInc")

# Lon vs MedInc
axes[2].scatter(df_min_max["Longitude"], df_min_max["MedInc"], 
                c=df_min_max["cluster"], cmap='viridis', alpha=0.6)
axes[2].set_title("Longitude vs MedInc")

plt.tight_layout()
plt.show()
```

---

### 3. SUGERENCIA: Silhouette Score

Lo importas pero nunca lo usas:

```python
from sklearn.metrics import silhouette_score  # ← Importado pero no usado
```

Añade:
```python
silhouette = silhouette_score(df_min_max.drop(columns=['cluster']), df_min_max['cluster'])
print(f"Silhouette Score: {silhouette:.4f}")
```

## Silhouette Score - Explicación

El **Silhouette Score** (coeficiente de silueta) es una métrica para evaluar la **calidad del clustering**. Responde a la pregunta: **¿Qué tan bien separados están los clusters?**

### ¿Qué mide?

Para cada punto, calcula dos cosas:

1. **a** = Distancia media a los demás puntos de **su mismo cluster** (cohesión interna)
2. **b** = Distancia media a los puntos del **cluster más cercano** (separación externa)

### ¿Cómo interpretar el resultado?

| Valor | Significado |
|-------|-------------|
| **+1** | Perfecto: clusters muy densos y bien separados |
| **+0.5 a +1** | Buena estructura de clusters |
| **+0.25 a +0.5** | Estructura débil, posible solapamiento |
| **~0** | Clusters solapados, puntos en la frontera |
| **Negativo** | Puntos asignados al cluster incorrecto |

### ¿Por qué usarlo?

1. **Complementa el método del codo**: El codo te dice cuántos clusters, el silhouette te dice si esos clusters tienen sentido

2. **Detecta problemas**: Si el silhouette es bajo, quizás:
   - Elegiste mal el número de clusters
   - Los datos no tienen estructura de clusters natural
   - Algunas variables no aportan a la separación

**En resumen**: El método del codo te ayuda a elegir **cuántos clusters**, el Silhouette Score te dice **qué tan buenos son esos clusters**. Son complementarios.

---

### 4. Falta matriz de confusión

Piensa que siempre que podamos hacerla la haremos, es un imprescindible para ver qué ha predicho nuestro modelo :) 

```python
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_test)
plt.figure(figsize=(8, 6))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues',
            xticklabels=[f'Cluster {i}' for i in range(6)],
            yticklabels=[f'Cluster {i}' for i in range(6)])
plt.xlabel('Predicción')
plt.ylabel('Real')
plt.title('Matriz de Confusión - Random Forest')
plt.show()
```

---

### 5. Falta classification_report

Veamos las métricas claramente completas con él.

```python
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred_test))
```

---

### 5. Sobrescribes la variable `model`

```python
model = KMeans(n_clusters=6, ...)  # K-Means
model.fit(df_min_max)
...
model = RandomForestClassifier(...)  # ← Sobrescribes el K-Means!
```

Mejor usa nombres diferentes:
```python
kmeans = KMeans(n_clusters=6, random_state=42)
kmeans.fit(df_min_max)
...
rf_model = RandomForestClassifier(random_state=42)
```

---

### 6. Falta guardar los modelos

```python
import pickle

# Guardar K-Means
with open('../models/kmeans_housing.pkl', 'wb') as f:
    pickle.dump(kmeans, f)

# Guardar Scaler
with open('../models/scaler_housing.pkl', 'wb') as f:
    pickle.dump(min_max_scaler, f)

# Guardar Random Forest
with open('../models/rf_housing.pkl', 'wb') as f:
    pickle.dump(rf_model, f)
```

---

### 7. Imports duplicados

Intenta dejar tus imports necesarios solo y limpios :) 

```python
import matplotlib.pyplot as plt  # ← Duplicado
from sklearn.cluster import KMeans  # ← Duplicado
```

---

### 8. Celda vacía

Tienes una celda vacía entre el elbow y el K-Means. Elimínala o añade contenido!!

---
